### PR TITLE
feat(namespace, idiomatic): add namespace support as per idiomatic maya guide-lines

### PR DESCRIPTION
**Pull request by mittachaitu**
Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

**What this PR does**:
This PR adds the support to build namespace and perform CRUD operations on the namespace by making use of client-go.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: pkg/kubernetes/namespace/v1alpha1/namespace.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	if len(name) == 0 {
+		b.errs = append(b.errs, errors.New("failed to build storageclass: missing storageclass name"))
+		return b
+	}
+	b.ns.object.Name = name
+	return b
+}
+
+// Build returns the namespace API instance
+func (b *Builder) Build() (*corev1.Namespace, error) {
```
Comment (sonasingh46) [inline pkg/kubernetes/namespace/v1alpha1/namespace.go]: `Build` should not return API object type.
Think of wrapper objects as the internal custom object with whom you want to associate behaviours.
It should return `*Namespace` type.

File: pkg/kubernetes/namespace/v1alpha1/namespace.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import (
+	errors "github.com/openebs/maya/pkg/errors/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Namespace is a wrapper over Namespace api
+// object. It provides build, validations and other common
+// logic to be used by various feature specific callers.
+type Namespace struct {
+	object *corev1.Namespace
```
Comment (sonasingh46) [inline pkg/kubernetes/namespace/v1alpha1/namespace.go]: You might need to export this. Correct me.

Comment (sonasingh46) [reply pkg/kubernetes/namespace/v1alpha1/namespace.go]: I think it is fine, if you do not export it.

Comment (sonasingh46) [reply pkg/kubernetes/namespace/v1alpha1/namespace.go]: But then -- it is good to have a getter function.

File: pkg/kubernetes/namespace/v1alpha1/namespace.go (hunk 1 above)
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/namespace.go]: I will agree with that. 
What would you say 
cc: @AmitKumarDas

File: pkg/kubernetes/namespace/v1alpha1/kubernetes_test.go [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+
+func fakeCreateFnErr(cli *kubernetes.Clientset, namespace *corev1.Namespace) (*corev1.Namespace, error) {
+	return nil, errors.New("failed to create Namespace")
+}
+
+func TestWithDefaultOptions(t *testing.T) {
+	tests := map[string]struct {
+		KubeClient *Kubeclient
+	}{
+		"When all are nil": {&Kubeclient{}},
```
Comment (imazik) [inline pkg/kubernetes/namespace/v1alpha1/kubernetes_test.go]: why we are using extra `{}`?

Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes_test.go]: initializing kubeclient with nil

File: pkg/kubernetes/namespace/v1alpha1/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete namespace {%s}", name)
+	}
+	return k.del(cli, name, deleteOpts)
+}
+
+// Create creates a namespace in specified namespace in kubernetes cluster
+func (k *Kubeclient) Create(namespace *corev1.Namespace) (*corev1.Namespace, error) {
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: No need but client-go call will take care of it

[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: call to create resource will throw error

File: pkg/kubernetes/namespace/v1alpha1/kubernetes.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+import (
+	"strings"
+
+	stringer "github.com/openebs/maya/pkg/apis/stringer/v1alpha1"
+	client "github.com/openebs/maya/pkg/kubernetes/client/v1alpha1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubernetes "k8s.io/client-go/kubernetes"
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/namespace/v1alpha1/kubernetes.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+// getClientsetFn is a typed function that
+// abstracts fetching of clientset
+type getClientsetFn func() (clientset *kubernetes.Clientset, err error)
+
+// getClientsetFromPathFn is a typed function that
+// abstracts fetching of clientset from kubeConfigPath
+type getClientsetForPathFn func(kubeConfigPath string) (clientset *kubernetes.Clientset, err error)
+
+// getFn is a typed function that abstracts
+// to get pod
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: Updated... thanks

File: pkg/kubernetes/namespace/v1alpha1/kubernetes.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"strings"
+
+	stringer "github.com/openebs/maya/pkg/apis/stringer/v1alpha1"
+	client "github.com/openebs/maya/pkg/kubernetes/client/v1alpha1"
+	"github.com/pkg/errors"
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/pod/v1alpha1/pod.go [hunk 8]
```diff
@@ -76,6 +76,32 @@ func IsRunning() Predicate {
...
+			}
+		}
+		return true
+	}
+}
+
+// HasLabel return true if provided lable
+// key and value are present in the the provided PodList
+// instance
+func (p *Pod) HasLabel(key, value string) bool {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/pod/v1alpha1/pod.go]: Expose a predicate as well.
Let this predicate use this method

File: pkg/kubernetes/namespace/v1alpha1/kubernetes.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+		return nil, errors.Wrapf(err, "failed to get namespace {%s}", name)
+	}
+	return k.get(cli, name, opts)
+}
+
+// Delete deletes a namespace instance from the
+// kubecrnetes cluster
+func (k *Kubeclient) Delete(name string, deleteOpts *metav1.DeleteOptions) error {
+	if strings.TrimSpace(name) == "" {
+		return errors.New("failed to delete namespace: missing namespace name")
```
Comment (AmitKumarDas) [inline pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: "failed to delete namespace: missing name"

Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: IMO good to have same log

File: pkg/kubernetes/namespace/v1alpha1/namespace.go (hunk 1 above)
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/namespace.go]: Done

File: pkg/kubernetes/namespace/v1alpha1/namespace.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+// NewBuilder returns new instance of Builder
+func NewBuilder() *Builder {
+	return &Builder{ns: &Namespace{object: &corev1.Namespace{}}}
+}
+
+// WithName sets the Name field of namespace with provided argument.
+func (b *Builder) WithName(name string) *Builder {
+	if len(name) == 0 {
+		b.errs = append(b.errs, errors.New("failed to build storageclass: missing storageclass name"))
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/namespace.go]: Updated

File: pkg/kubernetes/namespace/v1alpha1/kubernetes.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+// Create creates a namespace in specified namespace in kubernetes cluster
+func (k *Kubeclient) Create(namespace *corev1.Namespace) (*corev1.Namespace, error) {
+	if namespace == nil {
+		return nil, errors.New("failed to create namespace: namespace obj is nil")
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create namespace: %s", stringer.Yaml("namespace yaml", namespace))
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/namespace/v1alpha1/kubernetes.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete namespace {%s}", name)
+	}
+	return k.del(cli, name, deleteOpts)
+}
+
+// Create creates a namespace in specified namespace in kubernetes cluster
+func (k *Kubeclient) Create(namespace *corev1.Namespace) (*corev1.Namespace, error) {
+	if namespace == nil {
+		return nil, errors.New("failed to create namespace: namespace obj is nil")
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/pod/v1alpha1/pod.go (hunk 8 above)
Comment (mittachaitu) [reply pkg/kubernetes/pod/v1alpha1/pod.go]: Added another method

File: pkg/kubernetes/namespace/v1alpha1/kubernetes.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/namespace/v1alpha1/kubernetes.go]: Added check

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go [hunk 13]
```diff
@@ -218,9 +217,12 @@ func (k *Kubeclient) Delete(name string, deleteOpts *metav1.DeleteOptions) error
 
 // Create creates a pvc in specified namespace in kubernetes cluster
 func (k *Kubeclient) Create(pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	if pvc == nil {
+		return nil, errors.New("failed to create pvc: nil pvc object")
+	}
 	cli, err := k.getClientsetOrCached()
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to create pvc: %s", stringer.Yaml("persistent volume claim", pvc))
+		return nil, errors.Wrapf(err, "failed to create pvc")
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: "failed to create pvc {%s} in namespace {%s}", pvc.Name, pvc.Namespace

Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: Done

**Pull request merged**